enhancements/etcd/automated-defrag.md

```diff
@@ -0,0 +1,235 @@
+
+### Implementation Details/Notes/Constraints
+
+## etcd Storage Backend
+etcd's storage backend is powered by `bbolt` a B+Tree memory mapped key/value store with MVCC support. While MVCC is
+necessary to support fully concurrent transactions and the heart of the `Watch` functionality.
+MVCC poses a performance challenge for `bbolt` because the db can store many versions of the keys value. These
+versions can build up indefinitely and result in memory and disk performance issues. The solution is to regularly
+compact the key space.
+
+compaction is essentially a delete action in `bbolt`, where the prevision versions of a keys value are pruned.
+`bbolt` organizes data on pages and when a page is no longer used as the result of a db mutation it is called a
+free page and an accounting of all free pages are stored in a freelist for later reuse.
+
+Compaction does reduce the memory footprint by removing old keys but free pages can not be reclaimed back to disk
+without defragmentation.
+
+## Defrag Controller
+
+During the operator sync the controller will make a  call to the cluster API using the `MemberListRequest` RPC which 
+returns the current list of etcd members. We will then loop these members and for each dial the maintenance API using
+the `StatusRequest` RPC which will return detailed information on the backend status.
+
+```go
+type StatusResponse struct {
+    // dbSize is the size of the backend database physically allocated, in bytes, of the responding member.
+    DbSize int64 `protobuf:"varint,3,opt,name=dbSize,proto3" json:"dbSize,omitempty"`
+    [...] 
+    // dbSizeInUse is the size of the backend database logically in use, in bytes, of the responding member.
+    DbSizeInUse int64 `protobuf:"varint,9,opt,name=dbSizeInUse,proto3" json:"dbSizeInUse,omitempty"`
+}
+```
+
+From these two fields we can conclude the total fragmented space on that etcd members backend store. If the
+percentage of bytes of the total keyspace are over a defined threshold percentage. The controller will 
+utilize maintenance API again and issue a `DefragmentRequest` RPC.
+
+Defragmentation is I/O blocking thus important to only defrag a single etcd at a time. Additionally, to reduce
+the possibility of multiple leader elections we defragment the leader last. The amount of time that is required to
+defragment the store is directly related to the amount of free pages that exists. For this reason we feel it
+makes sense to continuously observe the clusters state and defrag frequently along the cadence of 
+compaction resulting in shorter intervals of disruption by keeping the number of pages limited and the size of the 
+database smaller.
+
+The criteria for defragmentation is defined as:
+
+- Cluster must be observed as `HighlyAvailableTopologyMode`.
+- The etcd cluster must report all members as healthy.
+- `minDefragBytes` this is a minimum database size in bytes before defragmentation should be considered.
```

Comment:
What is this value?

Reply:
I am thinking 100MB to reduce early defrag during install. Its possible this is not needed but I think 100MB would be reasonable.

Reply:
Sounds good, now I see the need to send this data back to telemetry, thanks!